src/bio/bio_buffer.c

```diff
@@ -637,6 +637,28 @@ rw_completion(void *cb_arg, int err)
 	D_ASSERT(xs_ctxt->bxc_blob_rw > 0);
 	xs_ctxt->bxc_blob_rw--;
 
+	/* Fault injection - Induce Write IO Error.
+	 */
+	if (DAOS_FAIL_CHECK(DAOS_NVME_BIO_WRITE_ERR)) {
+		D_ALLOC_PTR(mem);
+		mem->mem_err_type = MET_WRITE;
+		mem->mem_bs = xs_ctxt->bxc_blobstore;
+		mem->mem_tgt_id = xs_ctxt->bxc_tgt_id;
+		spdk_thread_send_msg(owner_thread(mem->mem_bs), bio_media_error,
+				     mem);
+	}
+
+	/* Fault injection - Induce Read IO Error.
+	 */
+	if (DAOS_FAIL_CHECK(DAOS_NVME_BIO_READ_ERR)) {
+		D_ALLOC_PTR(mem);
+		mem->mem_err_type = MET_READ;
+		mem->mem_bs = xs_ctxt->bxc_blobstore;
+		mem->mem_tgt_id = xs_ctxt->bxc_tgt_id;
+		spdk_thread_send_msg(owner_thread(mem->mem_bs), bio_media_error,
+				     mem);
+	}
```

Comment:
Minor: These two code snippets looks bit duplicated.  We can modify following real media error handling code to take error simulation into account.

Reply:
Actually initial code was having the below code to avoid the duplication. Thing is DAOS_FAIL_CHECK is one time check so once you use it for checking if it's Write/Read error, that can not be used second time for checking Error type. Test was not working as expected and I changed it to in separate function.
Let me know if you have better way to handle this code and I can change it.
```
	if (DAOS_FAIL_CHECK(DAOS_NVME_BIO_WRITE_ERR) ||
		DAOS_FAIL_CHECK(DAOS_NVME_BIO_READ_ERR)) {
		D_ALLOC_PTR(mem);
		mem->mem_err_type = (DAOS_FAIL_CHECK(DAOS_NVME_BIO_WRITE_ERR) ?
				     MET_WRITE : MET_READ);
		mem->mem_bs = xs_ctxt->bxc_blobstore;
		mem->mem_tgt_id = xs_ctxt->bxc_tgt_id;
		spdk_thread_send_msg(owner_thread(mem->mem_bs), bio_media_error,
				     mem);
	}
```

Reply:
Looks there are also few minor defects in this code, for example: 
- Should we simulate write error on read or simulate read error on write? I suppose our intention is to simulate write error on write and simulate read error on read, right?
- Should we fail the IO when simulating an IO error? I suppose we should.
- Should we report all NVMe IO errors if this 'biod' have multiple NVMe IO errors? I suppose we should report all NVMe IO errors.

What about this:
```
    ...
    if (biod->bd_update)
        err = DAOS_FAIL_CHECK(DAOS_NVME_BIO_WRITE_ERR) ? err : -DER_MISC;
    else
        err = DAOS_FAIL_CHECK(DAOS_NVME_BIO_READ_ERR) ? err : -DER_MISC;

    /* Return the error value of the first NVMe IO error */
    if (biod->bd_result == 0 && err != 0)
        biod->bd_result = err;

    /* Report all NVMe IO errors */
    if (err != 0) {
        D_ALLOC_PTR(mem);
        if (mem == NULL)
            goto skip_media_error;
        mem->mem_err_type = biod->bd_update ? MET_WRITE : MET_READ;
        mem->mem_bs = xs_ctxt->bxc_blobstore;
        mem->mem_tgt_id = xs_ctxt->bxc_tgt_id;
        spdk_thread_send_msg(owner_thread(mem->mem_bs), bio_media_error, mem);
    }
    ...
```

Reply:
That looks better, I will update the PR
Thanks for suggestion.